### PR TITLE
[rilbinder] Add support for android.hardware.radio@1.4. JB#54258

### PR DESCRIPTION
**Pull request by d-grigorev**

### CODE REVIEW

File: src/ril_binder_radio.c [hunk 1]
```diff
@@ -2338,6 +2606,62 @@ ril_binder_radio_decode_signal_strength_1_2(
...
+        grilio_encode_int32(out, strength->lte.rssnr);
+        grilio_encode_int32(out, strength->lte.cqi);
+        grilio_encode_int32(out, strength->lte.timingAdvance);
+
+        /* TD_SCDMA_SignalStrength */
+        grilio_encode_int32(out, strength->tdscdma.rscp);
+        return TRUE;
+    }
+    return FALSE;
+}
```
Comment (monich) [inline src/ril_binder_radio.c]: The above looks suspiciously similar to `ril_binder_radio_decode_signal_strength_1_2()` and even `ril_binder_radio_decode_signal_strength()`. Consider moving common code to one function. e.g.
```c
static
void
ril_binder_radio_decode_signal_strength_common(
    const RadioSignalStrengthGsm* gsm,
    const RadioSignalStrengthCdma* cdma,
    const RadioSignalStrengthEvdo* evdo,
    const RadioSignalStrengthLte* lte,
    const RadioSignalStrengthTdScdma* tdScdma,
    const RadioSignalStrengthWcdma* wcdma,
    GByteArray* out)
{
    /* GW_SignalStrength */
    if (wcdma && wcdma->signalStrength <= 31 && gsm->signalStrength > 31) {
        /*
         * Presumably, 3G signal. The wcdma field did't exist in RIL
         * socket times.
         *
         * Valid signal strength values for both 2G and 3G are (0-31, 99)
         * as defined in TS 27.007 8.5
         */
        grilio_encode_int32(out, wcdma->signalStrength);
        grilio_encode_int32(out, wcdma->bitErrorRate);
    } else {
        grilio_encode_int32(out, gsm->signalStrength);
        grilio_encode_int32(out, gsm->bitErrorRate);
    }

    /* CDMA_SignalStrength */
    grilio_encode_int32(out, cdma->dbm);
    grilio_encode_int32(out, cdma->ecio);

    /* EVDO_SignalStrength */
    grilio_encode_int32(out, evdo->dbm);
    grilio_encode_int32(out, evdo->ecio);
    grilio_encode_int32(out, evdo->signalNoiseRatio);

    /* LTE_SignalStrength_v8 */
    grilio_encode_int32(out, lte->signalStrength);
    grilio_encode_int32(out, lte->rsrp);
    grilio_encode_int32(out, lte->rsrq);
    grilio_encode_int32(out, lte->rssnr);
    grilio_encode_int32(out, lte->cqi);
    grilio_encode_int32(out, lte->timingAdvance);

    /* TD_SCDMA_SignalStrength */
    grilio_encode_int32(out, tdScdma->rscp);
}
```
and calling it like this:
```c
static
gboolean
ril_binder_radio_decode_signal_strength_1_4(
    GBinderReader* in,
    GByteArray* out)
{
    const RadioSignalStrength_1_4* strength = gbinder_reader_read_hidl_struct
        (in, RadioSignalStrength_1_4);

    if (strength) {
        RadioSignalStrengthTdScdma tdscdma;

        tdscdma.rscp = strength->tdscdma.rscp;
        ril_binder_radio_decode_signal_strength_common(&strength->gsm,
            &strength->cdma, &strength->evdo, &strength->lte,
            &tdscdma, &strength->wcdma.base, out);
        return TRUE;
    }
    return FALSE;
}
```
from all 3 places. Copy/paste is evil.

File: src/ril_binder_radio.c [hunk 2]
```diff
@@ -2840,6 +3178,90 @@ ril_binder_radio_decode_cell_info_list_1_2(
...
+                ril_binder_radio_decode_cell_info_header_1_4(out,
+                        cell, RADIO_CELL_INFO_WCDMA);
+                ril_binder_radio_decode_cell_info_wcdma(out,
+                        &cell->info.wcdma.cellIdentityWcdma.base,
+                        &cell->info.wcdma.signalStrengthWcdma.base);
+                break;
+            case RADIO_CELL_INFO_1_4_TD_SCDMA:
+                ril_binder_radio_decode_cell_info_header_1_4(out,
+                        cell, RADIO_CELL_INFO_TD_SCDMA);
+               ril_binder_radio_decode_cell_info_tdscdma(out,
```
Comment (monich) [inline src/ril_binder_radio.c]: nitpick: indentation slightly off

File: src/ril_binder_radio.c [hunk 3]
```diff
@@ -3573,6 +3995,67 @@ static const RilBinderRadioCall ril_binder_radio_calls_1_2[] = {
...
+        "setPreferredNetworkTypeBitmap_1_4"
+    },{
+        0,
+        0,
+        RADIO_RESP_GET_PREFERRED_NETWORK_TYPE_BITMAP,
+        NULL,
+        ril_binder_radio_decode_pref_network_type_bitmap,
+        "getPreferredNetworkTypeBitmap_1_4"
+    }
+
```
Comment (monich) [inline src/ril_binder_radio.c]: empty line

File: src/ril_binder_radio.c [hunk 4]
```diff
@@ -4249,18 +4751,39 @@ ril_binder_radio_arg_name(
         RIL_BINDER_DEFAULT_NAME);
 }
 
+static
+enum radio_interface
+ril_binder_radio_arg_interface(
+    GHashTable* args)
+{
+    const char *name = ril_binder_radio_arg_value(args,
+        RIL_BINDER_KEY_INTERFACE, RIL_BINDER_DEFAULT_INTERFACE);
```
Comment (monich) [inline src/ril_binder_radio.c]: There's no need to pass a non-NULL string as a default only to find it on the list and return a enum value (BTW why `enum radio_interface` and not `RADIO_INTERFACE`?). Get rid of RIL_BINDER_DEFAULT_INTERFACE, pass NULL, and return the default enum value if no interface is specified:
```c
static
RADIO_INTERFACE
ril_binder_radio_arg_interface(
    GHashTable* args)
{
    const char *name = ril_binder_radio_arg_value(args,
        RIL_BINDER_KEY_INTERFACE, NULL);

    if (name) {
        RADIO_INTERFACE i;

        for (i = RADIO_INTERFACE_1_0; i < RADIO_INTERFACE_COUNT; i++) {
            if (!g_strcmp0(name, ril_binder_radio_interface_name(i))) {
                return i;
            }
        }
    }
    return DEFAULT_INTERFACE;
}
```
There's no need to print anything here, especially as a WARNING, the selected value will by printed by `ril_binder_radio_init_base()`. And there's no need for `ril_binder_radio_interface_supported()` check, the logic in `ril_binder_radio_init_base()` should handle any value, supported or not.

File: src/ril_binder_radio.c [hunk 5]
```diff
@@ -283,6 +309,58 @@ ril_binder_radio_apn_types_for_profile(
...
+        return "radio@1.1";
+    case RADIO_INTERFACE_1_2:
+        return "radio@1.2";
+    case RADIO_INTERFACE_1_3:
+        return "radio@1.3";
+    case RADIO_INTERFACE_1_4:
+        return "radio@1.4";
+    default:
+        return NULL;
+    }
```
Comment (monich) [inline src/ril_binder_radio.c:340]: It's a good idea to not have a default. Doing this instead:
```c
static
const char*
ril_binder_radio_interface_name(
    RADIO_INTERFACE interface)
{
    switch (interface) {
    case RADIO_INTERFACE_1_0: return "radio@1.0";
    case RADIO_INTERFACE_1_1: return "radio@1.1";
    case RADIO_INTERFACE_1_2: return "radio@1.2";
    case RADIO_INTERFACE_1_3: return "radio@1.3";
    case RADIO_INTERFACE_1_4: return "radio@1.4";
    case RADIO_INTERFACE_COUNT:
        break;
    }
    return NULL;
}
```
Would produce "enumeration value not handled in switch" warning when a new enumeration value is introduced.

**Comment by d-grigorev** [comment]
I updated the code as you suggested. Thank you!

**Review by monich — APPROVED** [review]
Tested, haven't spotted any obvious regressions, LGTM.

**Pull request merged**